src/_zkapauthorizer/resource.py

```diff
@@ -214,110 +213,174 @@ def _setup_replication(self, request):
-        # it does, make sure it shows up in the log at least.
-        d.addErrback(partial(self._log.failure, "unhandled recovery failure"))
-
-        # _recover is responsible for generating the response.
-        return NOT_DONE_YET
+        else:
+            # we got another client, and they sent the same recovery
+            # capability, so be idempotent by acting the same as if this
+            # was the first client. That means sending this client all the
+            # status updates we've sent so far.
+            for update in self.sent_updates:
+                client.sendMessage(update)
+
+    def buildProtocol(self, addr) -> RecoverProtocol:
+        """
+        IFactory API
+        """
+        protocol = self.protocol()
+        protocol.factory = self
+        return protocol
 
     async def _recover(
         self,
-        request: IRequest,
         store: VoucherStore,
         cap: ReadonlyDirectoryURI,
-    ):
-        async def initiate(request, recoverer, downloader, cursor):
-            recovering = recoverer.recover(downloader, cursor)
-            recovering_d = Deferred.fromCoroutine(recovering)
-
-            # The only way to get to this point is if the store is empty and
-            # the recoverer appears to at least be willing to try to start.
-            # This is no guarantee that recovery will succeed but it is a
-            # guarantee that the *try* has begun.  Generate a response that
-            # reflects this.
-            request.setResponseCode(ACCEPTED)
-            request.finish()
-
-            # Really start the recovery attempt.
-            await recovering_d
-
+    ) -> None:
         try:
             # If these things succeed then we will have started recovery and
             # generated a response to the request.
-            downloader = self.get_downloader(cap)
+            downloader = self.get_downloader(cap.to_string().decode("ascii"))
             await store.call_if_empty(
-                partial(initiate, request, self.recoverer, downloader)
+                partial(
+                    self.recoverer.recover, downloader
+                )  # cursor added by call_if_empty
             )
         except NotEmpty:
             # If the database had anything in it, though, recovery will not be
             # attempted - and it will even fail quickly enough to be a
             # reasonable way to detect and report the conflict case.
-            request.setResponseCode(CONFLICT)
-            request.write(b"there is existing local state")
-            request.finish()
-        except:
-            # And if something else is broken, then who knows...  At least try
-            # to generate an error response.
-            self._log.failure("recovery setup failed")
-            request.setResponseCode(INTERNAL_SERVER_ERROR)
-            request.write(b"recovery setup failed")
-            request.finish()
+            raise ValueError("there is existing local state")
```

Comment:
Perhaps worth just letting `NotEmpty` through as well since now there is an exception handler wrapped around this whole coroutine.